Add WebGL ColorLayer Features.

gchoqueux opened the pull request:
## Description
* Set texture `ColorLayer` filtering.
* Add custom shader effect to `ColorLayer`.
* Cleaning of effect setting
* Refactoring chunk shader management
* add custom shader example 

![image](https://user-images.githubusercontent.com/11291849/117957341-44d0fa80-b31a-11eb-8cc0-183cafeffd9e.png)

gchoqueux committed:
feature(ColorLayer): add custom shader to ColorLayer.

gchoqueux committed:
refactor(ShaderChunk): introduce shader chunk manager class to instance ShaderChunk.

gmaillet commented on examples/customColorLayerEffect.html:
```diff
@@ -0,0 +1,140 @@
+                source: opensmSource,
+                type_effect: itowns.colorLayerEffects.customEffect,
+                amount_effect: 1.0,
+            });
```
this source it not very suitable for this example, is it possible to use a more classical cartographic layer with solid colors (like layers/JSONLayers/OPENSM.json

gmaillet commented on examples/customColorLayerEffect.html:
```diff
@@ -0,0 +1,140 @@
+
+                float value = edge(tex, step/resolution.x, step/resolution.y, cuv);
+
+                color = vec4(vec3(0.27, 0.4, 0.42) *  value * 0.8, value);
```
why these strange values rather than 
color = vec4(vec3(1.0, 0.0, 0.0) * value, value);

gmaillet commented on examples/customColorLayerEffect.html:
```diff
@@ -0,0 +1,140 @@
+                amount_effect: 1.0,
+            });
+
+            view.addLayer(opensmLayer);
```
it could be a good idea to add two layers in the view with the same source: one normal and one with the special effect.

gmaillet commented on examples/customColorLayerEffect.html:
```diff
@@ -0,0 +1,140 @@
+                },
+                source: opensmSource,
+                type_effect: itowns.colorLayerEffects.customEffect,
+                amount_effect: 1.0,
```
it could be a good idea to add magFilter and minFilter in these example

gmaillet commented on src/Layer/ColorLayer.js:
```diff
@@ -26,11 +27,13 @@ import Style from 'Core/Style';
+ * * `1`: light color to invisible effect.
+ * * `2`: white color to invisible effect.
+ * * `3`: custom shader effect (defined `ShaderChunk.customBodyColorLayer` and `ShaderChunk.customHeaderColorLayer`).
+ * @property {number} amount_effect - amount value used with effect applied on raster color.
```
"amount_effect": not very precise or explicit?
maybe some thing more generic like "effect_parameter" because it's not necessary a question of amount

gchoqueux committed:
examples: add custom shader effect example.

gchoqueux commented:
@gmaillet thanks, I have taken into account all your comments

mgermerie commented on src/Layer/ColorLayer.js:
```diff
@@ -50,6 +53,8 @@ class ColorLayer extends RasterLayer {
      * name.
      * @param {Source} [config.source] - Description and options of the source.
+     * @param {Source} [config.magFilter] - How the texture is sampled when a texel covers more than one pixel. [see](https://threejs.org/docs/?q=texture#api/en/textures/Texture.magFilter)
+     * @param {Source} [config.minFilter] - How the texture is sampled when a texel covers less than one pixel. [see](https://threejs.org/docs/?q=texture#api/en/textures/Texture.minFilter)
```
The `MagFilter` and `minFilter` values are supposed to be of type `number`, according to threejs doc.

mgermerie commented on src/Renderer/Shader/Chunk/color_layers_pars_fragment.glsl:
```diff
@@ -1,10 +1,13 @@
     float opacity;
 };
 
+#include <itowns/custom_header_colorLayer>
```
Could the custom header be included only if `Layer.effect_type` is worth `3` ?

gchoqueux replied:
the custom header always includes in preprocess

mgermerie commented on examples/vector_tile_raster_3d.html:
```diff
@@ -68,7 +68,8 @@
                 source: mvtSource,
-                fx: 2.5,
+                effect_type: itowns.colorLayerEffects.removeLightColor,
+                effect_parameter: 2.5,
```
I may be missing some point, but isn't the `fx: 2.5` previous value corresponding to a 'white color to invisible effect' ? If so, shouldn't the `effect_type` be `itowns.colorLayerEffects.removeWhiteColor` ?

gchoqueux replied:
yes,  it was an error in previous documentation

mgermerie commented on examples/customColorLayerEffect.html:
```diff
@@ -0,0 +1,141 @@
+            // Instanciate PlanarView
+            // By default itowns' tiles geometry have a "skirt" (ie they have a height),
+            // but in case of orthographic we don't need this feature, so disable it
+            var view = new itowns.PlanarView(viewerDiv, extent, { disableSkirt: true, maxSubdivisionLevel: 10,
```
Unless I'm wrong : since the camera is orthographic, setting `disableSkirt` to `true` or `false` does not have a visual effect.
Is there another effect I am missing with this parameter, like a performance effect or something ?

gchoqueux replied:
Skirt removes terrain visual artefact when there's a DEM.

gchoqueux committed:
feature(ColorLayer): add option to filtering textures Layer.

gchoqueux committed:
refactor(ColorLayer): add effect type ColorLayer parameter.

DEPRECATED:
ColorLayer.fx is deprecated, use type_effect and amount_effect instead.
